### DISCUSSION TITLE
Mutable Cache with Background Refreshes

### DISCUSSION THREAD

**Discussion by r39132** [ben-manes/caffeine#520] [Q&A] [answered]
Hi,
I have the following requirements for a local cache. I'm trying to understand if caffeine is a good fit. We have a high-fidelity data streams-as-a-service business with each microservice in the data plane running on Spring Boot. We currently use caffeine. One issue we have run into is that when a key is missing (either due to expiry or microservice cold-start), our data streams pause awaiting the cache entry to be load. This loading causes us to fetch data via our "cache wrapper microservice" which hits redis. The cache wrapper service does expose batch fetch APIs, so we can load and reload the entire local cache as needed, as long as it happens in the background and at a fixed cadence. 

A wrinkle, on cache miss, we do want to block the data stream pipeline to fetch the latest data from the central cache via the cache wrapper. 

Refer to details below: 

<img width="1362" alt="Screen Shot 2021-03-23 at 4 49 18 PM" src="https://user-images.githubusercontent.com/581734/112233470-bda7b600-8bf7-11eb-90ac-a5ec7a35753f.png">

I mentioned that I want key to stick around for an "infinite amount of time", by which I meant the lifetime of the app, but I don't mind replacement as long as expired keys can be refreshed in the background and on expiry. I don't want any pauses in data streaming... the caches are needed for stream processing.

I was originally planning on writing my own cache manager with each cache implemented as a loading cache, but then I noticed that it's immutable. A long time ago, I wrote a lot of the core caching functionality at Netflix -- I used soft and weak references and built a lot of logic around LInkedHashMap. I don't want to do that all again if there is a better way :-)

**Comment by ben-manes** [comment, marked as answer]
> A wrinkle, on cache miss, we do want to block the data stream pipeline to fetch the latest data from the central cache via the cache wrapper.

You can avoid blocking by using `getIfPresent` to obtain the entry or null. The `asMap()` view might also be helpful.

Our refresh support is intended to be used with expiration, e.g. to allow active content to stay fresh and inactive content to fade away. A full reload is best suited by a scheduled, period task.

The `cache.get` is an alias for `Map.computeIfAbsent`, so you can use that only when helpful for blocking calls.

> I was originally planning on writing my own cache manager with each cache implemented as a loading cache, but then I noticed that it's immutable.

I'm not sure what you mean by this, but I am also not very familiar with Spring Cache. We don't provide the cache manager and related framework, as it would duplicate efforts of the platform (spring, jcache, custom). It may be that Spring's needs to know all of the caches upfront.

> A long time ago, I wrote a lot of the core caching functionality at Netflix -- I used soft and weak references and built a lot of logic around LInkedHashMap. I don't want to do that all again if there is a better way :-)

Caffeine is basically a fancy `ConcurrentHashMap` with various bounding options. If you think of it in those terms, like you did when writing your own, then you can choose the preferred features. It doesn't offer frameworky things beyond that, but rather provides the core data structure to base your own on top of.

**Reply by r39132** [reply]
I'm trying not to use Spring and get to the core data structures actually.

What is the purpose of a loading cache? Perhaps, put another way, which interface and/or implementation is closest to what I am trying to achieve above?

**Reply by ben-manes** [reply]
In Guava's time, `LoadingCache` was primarily to promote loading through the cache rather than the popular, racy `get-load-put` idiom. The more common approach leads to cache stampedes and performance degradations. We wanted to nudge users towards delegating loading to the cache, as well as simpler code by pulling complex logic into a cache loader class.

Almost all of the functionality is available in `Cache`, such as bulk loads. The only unique feature that the attached cache loader adds is the ability to use `refreshAfterWrite`.

Guava/Caffeine refresh is intended to hide latency for a popular, but soon to expire entry. When a popular entry expires, like a configuration value, then all callers suffer a penalty waiting for it to be loaded. If the refresh interval is shorter than the expiration interval, then a read will trigger it to reloaded asynchronously and its duration reset. On the other hand, if the entry is not accessed then expiration will kick in and evict, thereby bounding the cache. Similarly you can manually call `LoadingCache.refresh(key)` to reload asynchronously while the cache serves stale content, e.g. if a remote cache is updated and an invalidation is broadcasted to your local caches for a given key.

**Reply by ben-manes** [reply]
In your case, the `size limit` and `blocking on-demand` are serviced by `Cache`. The `non-blocking refresh` is best done how you describe, since it is a batch call across all keys that inserts / updates the existing mapping and not a per-key reload to hide latency due to expiration.

**Reply by r39132** [reply]
How do I implement the load-entry-on-cache-miss functionality? I'd essentially like to register a function that fetch an entry from the central Redis cache on local cache miss. In my case, the reason for a cache miss is that the local cache doesn't have the entry. Once an entry has been loaded once into my local cache, a background refresher will reload the entire cache as a fixed cadence. My caches are not persistent, so they don't survive JVM restarts -- it's not a requirement that they do, so it's by design.  

I currently need to solve the cold-start cache problem, which I would like to solve in a lazy way -- on local cache miss, look entry up from Redis.

Below is the method I use to build empty caches. How do I register a method that will execute a load entry function and block all get() callers in the process until the load completes?

<img width="782" alt="Screen Shot 2021-03-26 at 12 08 59 AM" src="https://user-images.githubusercontent.com/581734/112595327-80921e00-8dc7-11eb-9744-b9367299a72d.png">

**Reply by r39132** [reply]
One example I found via [this tutorial](https://www.baeldung.com/java-caching-caffeine) is 
```
dataObject = cache
  .get(key, k -> DataObject.get("Data for A"));
```

**Reply by r39132** [reply]
Here is what I am thinking.

<img width="771" alt="Screen Shot 2021-03-26 at 12 25 22 AM" src="https://user-images.githubusercontent.com/581734/112596994-cc45c700-8dc9-11eb-8628-3bddfbe8bbb8.png">

**Reply by ben-manes** [reply]
That is exactly right if you want a blocking call to load on a miss and avoid cache stampedes. This will internally perform a `Map.get` and, if the entry is absent or invalid, then perform a `Map.compute` to load it.

Earlier you stated that you wanted to avoid blocking reads for the stream operations. If it is tolerable to miss data, then you can use `getIfPresent` which delegates to `Map.get` only.

We also offer an `AsyncCache` which you may find useful. This is equivalent to `Map<K, CompletableFuture<V>>` where the entry is visible in the map once the in-flight future has been inserted. That loading is then performed on a background executor, which means if your stream operations can continue doing work you may be able to hide some of the blocking latencies. It also offers an improved `getAll` if you can batch up work (perhaps by coalescing individual keys) which is performed atomically to protect against cache stampedes (unlike `Cache.getAll` which is forced to be non-blocking due to `ConcurrentHashMap` limitations).

It may also be worth noting that the caches will never store `null` or exceptional conditions. This may be obvious since it is a fancier `ConcurrentHashMap`. Sometimes negative caching can be beneficial and you can translate a failure to a success by wrapping the value in an `Optional`. If you wanted a different expiration time for that then the `expireAfter(expiry)` custom policy could evaluate to give it a shorter duration. Similarly it might be opted out of the capacity limit by using a custom `Weigher` that returns zero.

**Reply by r39132** [reply]
Hi Ben,
I am trying to test the refreshAfterWrite and expireAfterWrite functionality in the LoadingCache, but it is not working.

Review the following classes:


**1. This Utils class defines how the caches are created:**

<img width="826" alt="Screen Shot 2021-04-06 at 5 37 04 PM" src="https://user-images.githubusercontent.com/581734/113794160-c8397380-96fe-11eb-9dd7-4e352b2dda79.png">


**2. Here is the CacheLoader I have defined:**

<img width="743" alt="Screen Shot 2021-04-06 at 5 39 41 PM" src="https://user-images.githubusercontent.com/581734/113794319-1e0e1b80-96ff-11eb-9c5c-db520752431a.png">


**3. Here are 2 Tests that are failing:**

<img width="583" alt="Screen Shot 2021-04-06 at 5 43 04 PM" src="https://user-images.githubusercontent.com/581734/113794522-9b399080-96ff-11eb-8305-3baa42844da2.png">

<img width="600" alt="Screen Shot 2021-04-06 at 5 42 04 PM" src="https://user-images.githubusercontent.com/581734/113794508-91179200-96ff-11eb-82fe-8b13f2cd06c9.png">

**4. Here are Exception stacks:**

<img width="944" alt="Screen Shot 2021-04-06 at 5 48 39 PM" src="https://user-images.githubusercontent.com/581734/113794868-624deb80-9700-11eb-9d76-983607967321.png">

<img width="1033" alt="Screen Shot 2021-04-06 at 5 48 28 PM" src="https://user-images.githubusercontent.com/581734/113794874-64b04580-9700-11eb-9b58-c7915b71530b.png">

**Reply by ben-manes** [reply]
I recommend reading this [wiki page](https://github.com/ben-manes/caffeine/wiki/Testing) for improving your tests.

A small tip is to use replace `Thread.sleep` with a `Ticker`, a logical clock. You can use Guava's testlib which has a `FakeTicker` utility. Then you can advance time.

A refresh occurs asynchronously, so you are subject to race conditions which can test failures. If you specify a direct, same-thread executor (e.g. `Runnable::run`) then the work will be performed by the caller.

To debug, you might add a log message in `CacheLoader#reload`. This by default calls `load(key)`, but you can specify something else. This would help you see what the cache is doing.

**Reply by r39132** [reply]
Ben.
**1. I have extended my code to support running in the calling thread and to use tickers.**  Note that I use com.google.common.base.Ticker in this class so that the Guava Testlib's Fake Ticker is compatible.

<img width="937" alt="Screen Shot 2021-04-06 at 6 49 36 PM" src="https://user-images.githubusercontent.com/581734/113798881-dbe9d780-9708-11eb-9290-047f010d1340.png">

**2. I have changed my test to use the Fake Ticker and I have added log calls ** 
<img width="958" alt="Screen Shot 2021-04-06 at 6 47 15 PM" src="https://user-images.githubusercontent.com/581734/113798806-bbba1880-9708-11eb-87e8-78821557323e.png">

**3. See the output. Some thing is strange. The loaded key is not visible in the next get() call. The stale value is still returned.  I get the same result whether I run in the same thread executor or not.** 
<img width="1680" alt="Screen Shot 2021-04-06 at 6 48 37 PM" src="https://user-images.githubusercontent.com/581734/113798815-beb50900-9708-11eb-9ba9-b684d2fcb142.png">


I expect to get the Object with samplingRate == 12, but I still get 11.

**Reply by ben-manes** [reply]
I am free if you would like to talk on a Zoom, as to be honest it's a bit hard to debug code over screenshots. :)

**Reply by r39132** [reply]
Note : My refreshAfterExpiry() works after I fixed the sleep bug you caught. I will move it to ticker as well, but the test works.

<img width="911" alt="Screen Shot 2021-04-06 at 6 57 06 PM" src="https://user-images.githubusercontent.com/581734/113799473-fe302500-9709-11eb-9368-adf7d61c8b5c.png">


<img width="1673" alt="Screen Shot 2021-04-06 at 6 59 34 PM" src="https://user-images.githubusercontent.com/581734/113799611-40596680-970a-11eb-8a51-3f5113708d0b.png">

**Comment by r39132** [comment]
**RefreshAfterWrite**
Why is this not working? I have set the refresh interval at 5 seconds and wait 10 seconds for the entry to refresh, but still get the same value for one of the fields in the object -- you will notice that I expect to get 11, 12, 13.... on subsequent calls. 

**RefreshAfterExpiry**
Similar question here.. on key expiry, shouldn't I expect to get different objects.

Also, I do the load() function execute one time, but only on the first get. I don't see it get called when the entry should have expired or become stale. 

```17:35:43.197 [Test worker] INFO io.datazoom.connector.base.cache.ConnectorConfigCacheLoader - Load called for key = name returned CMD = ConnectorMetadata(mappingId=name, dzConnectorId=null, customerCode=null, apiUrl=null, active=true, dzInternal=false, samplingRate=11, metadata={}) ```

**Reply by ben-manes** [reply]
Would you be able to generalize the code so that I can run it through a debugger?

FYI - `Thread.sleep` is in milliseconds, but your refreshIntervalSeconds is only `5`, not `5000`. You might be suffering from races.

**Reply by r39132** [reply]
Ah.. good catch on the thread.sleep :-)

**Comment by r39132** [comment]
Ben,
Our caffeine-based cache is now in production! I have observability (monitoring and alerting) via AWS Cloudwatch enabled. I am using caffeine's cache stats. 

I have a question about the unit of the avg_cache_load_penalty. Is it in milliseconds or microseconds? If in milliseconds, the numbers I am seeing don't make sense. For example, 30,000,000 in microseconds would be 30s, but in milliseconds would be 8.3 hours.

Also, how is this stat computed within caffeine? I am using the Loading cache with a 5m refresh interval and 3 hour TTL for my cache entries. 

<img width="1027" alt="Screen Shot 2021-04-15 at 2 10 08 PM" src="https://user-images.githubusercontent.com/581734/114938724-50e99b00-9df4-11eb-913a-10a95dc2e6cf.png">

**Reply by ben-manes** [reply]
That's wonderful!

That's a good JavaDoc fix. The time measurement is in nanoseconds, per `totalLoadTime`. The average doesn't state that, unfortunately. The calculation is `totalLoadTime / (loadSuccessCount + loadFailureCount)`.

We call `Ticker.read()` prior to invoking the loading function, call it again when the function returns the value, and update the `StatsCounter`. This means it is the time that it took to acquire the entry lock + run your function. Every cache load, whether explicit or a refresh, is captured by these metrics.

**Reply by r39132** [reply]
Great! Super useful. Thanks for the clarification @ben-manes. 

Nanoseconds makes sense. 

Our load function calls a various APIs in a microservice called `Cache-Wrapper`, which adds some light business logic, transparent redis failover logic, and observability around all redis (central cache) operations. So, the numbers make sense. The p95 for the API calls is a few single millis, which is less than the 30 mills we see for `avg_cache_load_penalty`.


<img width="1516" alt="Screen Shot 2021-04-15 at 2 42 51 PM" src="https://user-images.githubusercontent.com/581734/114943359-b93b7b00-9dfa-11eb-908c-7f3ad8831946.png">
